docs/high-availability.md

```diff
@@ -0,0 +1,177 @@
+```
+
+On each node, copy the [etcd.manifest](high-availability/etcd.manifest) file into ```/etc/kubernetes/manifests/etcd.manifest```
+
+The kubelet on each node actively monitors the contents of that directory, and it will create an instance of the ```etcd```
+server from the definition of the pod specified in ```etcd.manifest```.
+
+Note that you should substitute the token URL you got above for ```${DISCOVERY_TOKEN}``` on all three machines,
+and you should substitute a different name (e.g. ```node-1```) for ${NODE_NAME} and the correct IP address
+for ```${NODE_IP}``` on each machine.
+
+
+### Validating your cluster
+Once you copy this into all three nodes, you should have a clustered etcd set up.  You can validate with
+```
+etcdctl member list
+```
+
+and
+
+```
+etcdctl cluster-health
+```
+
+You can also validate that this is working with ```etcdctl set foo bar``` on one node, and ```etcd get foo```
+on a different node.
+
+## Replicated API Servers
+Once you have replicated etcd set up correctly, we will also install the apiserver using the kubelet.
+
+### Installing configuration files
+First you need to create the initial log file, so that Docker mounts a file instead of a directory:
+
+```
+touch /var/log/kube-apiserver.log
+```
+
+Next, you need to create a ```/srv/kubernetes/``` directory on each node.  This directory includes:
+   * basic_auth.csv  - basic auth user and password
+   * ca.crt - Certificate Authority cert
+   * known_tokens.csv - tokens that entities (e.g. the kubelet) can use to talk to the apiserver
+   * kubecfg.crt - Client certificate, public key
+   * kubecfg.key - Client certificate, private key
+   * server.cert - Server certificate, public key
+   * server.key - Server certificate, private key
+
+The easiest way to create this directory, may be to copy it from the master node of a working cluster, or you can manually generate these files yourself.
+
+### Starting the API Server
+Once these files exist, copy the [kube-apiserver.manifest](high-availability/kube-apiserver.manifest) into ```/etc/kubernetes/manifests/``` on each master node.
+
+The kubelet monitors this directory, and will automatically create an instance of the ```kube-apiserver``` container using the pod definition specified
+in the file.
+
+### Load balancing.
+At this point, you should have 3 apiservers all working correctly.  If you set up a network load balancer, you should
+be able to access your cluster via that load balancer, and see traffic balancing between the apiserver instances.  Setting
+up a load balancer will depend on the specifics of your load balancer, for example instructions for the Google Cloud
+Platform can be found [here](https://cloud.google.com/compute/docs/load-balancing/)
+
+Note, if you are using authentication, you may need to regenerate your certificat to include the IP address of the balancer,
+in addition to the IP addresses of the individual nodes.
+
+## Master-elected components
```

Comment:
this is a really slick approach. using a kubelet managed pod that manages the active scheduler/controller-manager by adjusting kubelet config and coordinating through etcd.

due to delays, you can have a situation where one kubelet is about to kill the active component while another kubelet has already started up the new active component. in this case you have two active components. what guarantees are in place to avoid state corruption or incorrect behavior (e.g. over/under provisioning w/ repl controllers, or double pod assignment by scheds)?